watch/source.go

```diff
@@ -53,20 +58,33 @@ func NewSource(poll SourcePollFn, logger xlog.Logger) Source {
 type source struct {
 	sync.RWMutex
 
-	poll   SourcePollFn
-	w      Watchable
-	logger xlog.Logger
-	closed bool
+	poll        SourcePollFn
+	w           Watchable
+	closed      bool
+	initialized bool
+	logger      xlog.Logger
 }
 
 func (s *source) run() {
 	for !s.isClosed() {
 		data, err := s.poll()
+		if err == ErrSourceClosed {
```

Comment:
This error is expected from the SourcePollFn, the impl of SourcePollFn could return this error to tell Source to stop polling it.